command/apply.go

```diff
@@ -22,105 +21,169 @@ type ApplyCommand struct {
+	view.ResourceCount()
+	if !c.Destroy && op.State != nil {
+		view.Outputs(op.State.RootModule().OutputValues)
+	}
+
+	view.Diagnostics(diags)
+
+	if diags.HasErrors() {
 		return 1
 	}
 
-	// Try to load plan if path is specified
+	return 0
+}
+
+func (c *ApplyCommand) LoadPlanFile(path string) (*planfile.Reader, tfdiags.Diagnostics) {
 	var planFile *planfile.Reader
-	if planPath != "" {
-		planFile, err = c.PlanFile(planPath)
+	var diags tfdiags.Diagnostics
+
+	// Try to load plan if path is specified
+	if path != "" {
+		var err error
+		planFile, err = c.PlanFile(path)
 		if err != nil {
-			c.Ui.Error(err.Error())
-			return 1
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Failed to load %q as a plan file", path),
+				fmt.Sprintf("Error: %s", err),
+			))
+			return nil, diags
 		}
 
 		// If the path doesn't look like a plan, both planFile and err will be
 		// nil. In that case, the user is probably trying to use the positional
 		// argument to specify a configuration path. Point them at -chdir.
 		if planFile == nil {
-			c.Ui.Error(fmt.Sprintf("Failed to load %q as a plan file. Did you mean to use -chdir?", planPath))
-			return 1
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Failed to load %q as a plan file", path),
+				"The specified path is a directory, not a plan file. You can use the global -chdir flag to use this directory as the configuration root.",
+			))
+			return nil, diags
 		}
 
 		// If we successfully loaded a plan but this is a destroy operation,
 		// explain that this is not supported.
 		if c.Destroy {
-			c.Ui.Error("Destroy can't be called with a plan file.")
-			return 1
-		}
-	}
-	if planFile != nil {
-		// Reset the config path for backend loading
-		configPath = ""
-
-		if !c.variableArgs.Empty() {
 			diags = diags.Append(tfdiags.Sourceless(
 				tfdiags.Error,
-				"Can't set variables when applying a saved plan",
-				"The -var and -var-file options cannot be used when applying a saved plan file, because a saved plan includes the variable values that were set when it was created.",
+				"Destroy can't be called with a plan file",
+				fmt.Sprintf("If this plan was created using plan -destroy, apply it using:\n  terraform apply %q", path),
 			))
-			c.showDiagnostics(diags)
-			return 1
+			return nil, diags
 		}
 	}
 
-	// Set up our count hook that keeps track of resource changes
-	countHook := new(CountHook)
+	return planFile, diags
+}
+
+func (c *ApplyCommand) PrepareBackend(planFile *planfile.Reader, args *arguments.State) (backend.Enhanced, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	// FIXME: we need to apply the state arguments to the meta object here
+	// because they are later used when initializing the backend. Carving a
+	// path to pass these arguments to the functions that need them is
+	// difficult but would make their use easier to understand.
+	c.Meta.applyStateArguments(args)
```

Comment:
FIXME! This one is so involved that I couldn't even come close to tackling it in this PR. The state file paths and lock settings are used in many places. I think it should be feasible to control them directly from the individual commands, though.